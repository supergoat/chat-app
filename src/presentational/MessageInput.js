// @flow
import React, { Component } from 'react';
import {
  Image,
  Keyboard,
  KeyboardAvoidingView,
  Platform,
  TextInput,
  TouchableOpacity,
  View
} from 'react-native';
import CAColors from "../common/CAColors";
import CAStyleSheet from "../common/CAStyleSheet";
import { isIphoneX } from '../common/CAHelpers';
import sendMessageIcon from '../assets/img/send_message.png';

/* Constants
============================================================================= */
const INPUT_HEIGHT_ANDROID = 42;
const INPUT_HEIGHT_IOS = 35;
const IPHONE_X_HEIGHT_ADJUSTMENT = 30;

/** ============================================================================
<MessageInput />
--------------------------------------------------------------------------------
Used to send messages
@param { function } onSendMessage - Function called when viewer sends message
@property { string } message - The message being send
============================================================================= */
type Props = {
  onSendMessage: (string) => void
}
type State = {
  message: string
}
class MessageInput extends Component<Props, State> {

  // TODO: remove padding bottom for iphone X when keyboard is open
  static __cards__: Function

  constructor() {
    super();
    this.state = { message: '' };
  }

  onSendMessage = () => {
    const message = this.state.message
    this.setState({ message: '' });
  }

  render() {
    return (
      <KeyboardAvoidingView
        style={styles.container}
        behavior={Platform.OS === 'ios' ? 'padding' : 'default'}
        enabled
      >
        <View style={[
          styles.messageInput,
          { paddingBottom: isIphoneX() ? IPHONE_X_HEIGHT_ADJUSTMENT : 0 }
        ]}>
          <TextInput
            multiline
            maxHeight={100}
            placeholder='Type a message'
            value={this.state.message}
            onChangeText={message => this.setState({ message })}
            style={styles.textInput}
          />
          <TouchableOpacity
            activeOpacity={0.8}
            style={styles.sendMessage}
            onPress={this.onSendMessage}
          >
            <Image style={styles.sendMessageIcon} source={sendMessageIcon} />
          </TouchableOpacity>
        </View>
      </KeyboardAvoidingView>
    )
  }
}

/* StyleSheet
============================================================================= */
const styles = CAStyleSheet.create({
  container: {
    ios: {
      backgroundColor: CAColors.alabaster,
      borderTopWidth: 1,
      borderTopColor: CAColors.gallery,
    }
  },
  messageInput: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    margin: 8
  },
  textInput: {
    flex: 1,
    marginRight: 8,
    paddingTop: 8,
    paddingBottom: 8,
    paddingHorizontal: 15,
    backgroundColor: CAColors.white,
    fontSize: 16,
    ios: {
      borderWidth: 1.5,
      borderRadius: INPUT_HEIGHT_IOS / 2,
      borderColor: CAColors.gallery,
      minHeight: INPUT_HEIGHT_IOS
    },
    android: {
      borderWidth: 0.5,
      borderRadius: INPUT_HEIGHT_ANDROID / 2,
      borderColor: CAColors.riverBed,
      minHeight: INPUT_HEIGHT_ANDROID
    }
  },
  sendMessage: {
    aspectRatio: 1,
    ios: {
      width: INPUT_HEIGHT_IOS,
      borderRadius: INPUT_HEIGHT_IOS / 2,
      padding: 10,
      backgroundColor: CAColors.mariner,
    },
    android: {
      padding: 12,
      width: INPUT_HEIGHT_ANDROID,
      borderRadius: INPUT_HEIGHT_ANDROID / 2,
      backgroundColor: CAColors.riverBed,
    }
  },
  sendMessageIcon: {
    width: '100%',
    height: '100%'
  }
});

/* Playground Cards - the different states of MessageInput
============================================================================= */
// MessageInput.__cards__ = define => {
//   define("Simple", () => <MessageInput />);
// };


/* Export
============================================================================= */
export default MessageInput;
