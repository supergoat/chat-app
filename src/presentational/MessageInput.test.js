// @flow
import React from 'react';
import MessageInput from './MessageInput';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

describe('MessageInput', () => {

  it('renders correctly', () => {
    const tree = renderer.create(<MessageInput />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('sets the KeyboardAvoidingView behaviour to padding when iOS', () => {
    const component = shallow(<MessageInput />);
    const behavior = component.prop('behavior');
    expect(behavior).toEqual('padding');
  });

  it('sets the KeyboardAvoidingView behaviour to default when android', () => {
    // Set platform to android
    jest.mock('Platform', () => ({
      OS: 'android',
    }));

    const component = shallow(<MessageInput />);
    const behavior = component.prop('behavior');
    expect(behavior).toEqual('default');
  });

  it('sets the state to the message entered in the textInput', () => {
      const component = shallow(<MessageInput />);
      const textInput = component.find('TextInput');
      textInput.simulate('ChangeText', 'hello');
      expect(component.state('message')).toEqual('hello');
  });

  it('sets the state to the message entered in the textInput', () => {
      const component = shallow(<MessageInput />);
      const textInput = component.find('TextInput');
      textInput.simulate('ChangeText', 'hello');
      expect(component.state('message')).toEqual('hello');
  });

  it('clears the state message and calls onSendMessage from props', () => {
      const component = shallow(<MessageInput />);
      const sendMessage = component.find('TouchableOpacity');
      component.setState({ message: 'hello'})
      sendMessage.simulate('Press');
      expect(component.state('message')).toEqual('');
  });

  it('sets a paddingBottom to messageInput when the phone is iphone X', () => {
    const CAHelpers = require('../common/CAHelpers');

    // Mock isIphoneX to return true
    CAHelpers.isIphoneX = jest.fn(() => true);

    const tree = renderer.create(<MessageInput />).toJSON();
    expect(tree).toMatchSnapshot();
  });

});
