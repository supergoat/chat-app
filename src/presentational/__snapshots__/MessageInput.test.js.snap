// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageInput renders correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#F9F9F9",
        "borderTopColor": "#EEEEEE",
        "borderTopWidth": 1,
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "margin": 8,
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      maxHeight={100}
      multiline={true}
      onChangeText={[Function]}
      placeholder="Type a message"
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "borderColor": "#EEEEEE",
          "borderRadius": 17.5,
          "borderWidth": 1.5,
          "flex": 1,
          "fontSize": 16,
          "marginRight": 8,
          "minHeight": 35,
          "paddingBottom": 8,
          "paddingHorizontal": 15,
          "paddingTop": 8,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "aspectRatio": 1,
          "backgroundColor": "#2277D8",
          "borderRadius": 17.5,
          "opacity": 1,
          "padding": 10,
          "width": 35,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/assets/img/send_message.png",
          }
        }
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`MessageInput sets a paddingBottom to messageInput when the phone is iphone X 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "backgroundColor": "#F9F9F9",
      "borderTopColor": "#EEEEEE",
      "borderTopWidth": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "margin": 8,
        },
        Object {
          "paddingBottom": 30,
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      maxHeight={100}
      multiline={true}
      onChangeText={[Function]}
      placeholder="Type a message"
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "borderColor": "#EEEEEE",
          "borderRadius": 17.5,
          "borderWidth": 1.5,
          "flex": 1,
          "fontSize": 16,
          "marginRight": 8,
          "minHeight": 35,
          "paddingBottom": 8,
          "paddingHorizontal": 15,
          "paddingTop": 8,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "aspectRatio": 1,
          "backgroundColor": "#2277D8",
          "borderRadius": 17.5,
          "opacity": 1,
          "padding": 10,
          "width": 35,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/assets/img/send_message.png",
          }
        }
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
      />
    </View>
  </View>
</View>
`;
